### DISCUSSION TITLE
There is a bug in the implementation of the Auth class

### DISCUSSION THREAD

**Discussion by nsano-rururu** [jertel/elastalert2#449] [Ideas]
Currently, if there is a value in the argument aws_region or the environment variable AWS_DEFAULT_REGION, it will proceed to the aws process, but if the process is executed without the profile_name value, an error will occur. I think we need to add a check for the value of profile_name as well.
https://github.com/jertel/elastalert2/blob/master/elastalert/auth.py#L53

current

```
        if not aws_region and not os.environ.get('AWS_DEFAULT_REGION'):
            return None

        session = boto3.session.Session(profile_name=profile_name, region_name=aws_region)

        return RefeshableAWSRequestsAuth(
            refreshable_credential=session.get_credentials(),
            aws_host=host,
            aws_region=session.region_name,
            aws_service='es')
```

fix idea

```
        if not aws_region and not os.environ.get('AWS_DEFAULT_REGION'):
            return None

        if not profile_name:
            return None

        session = boto3.session.Session(profile_name=profile_name, region_name=aws_region)

        return RefeshableAWSRequestsAuth(
            refreshable_credential=session.get_credentials(),
            aws_host=host,
            aws_region=session.region_name,
            aws_service='es')
```

**Comment by jertel** [comment]
According to the [documentation](https://elastalert2.readthedocs.io/en/latest/recipes/signing_requests.html?highlight=profile) the signing auth process attempts the following:

1. If user/pass provided, use standard http auth.
2. If region is provided but not profile, assume using an ["Instance" profile](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2.html) (which does not require the profile_name param.
3. If region and profile are both provided assume using an AWS profile (via the client's `.aws/` directory)

If that documentation is accurate then your proposed change would potentially block #2 from working. I've not tried this, and I don't use this configuration, so I am unable to validate this.

**Reply by nsano-rururu** [reply]
In the AWS Workspaces environment I'm using, the environment variable "AWS_DEFAULT_REGION" has a value set, so direct installation of elastalert2 is not supported.
When checking the operation locally, modify the source as follows and do "return None" and python setup.py install so that you can not proceed further.

```
        if not aws_region and not os.environ.get('AWS_DEFAULT_REGION'):
            return None

        # add code
        return None

        session = boto3.session.Session(profile_name=profile_name, region_name=aws_region)

        return RefeshableAWSRequestsAuth(
            refreshable_credential=session.get_credentials(),
            aws_host=host,
            aws_region=session.region_name,
            aws_service='es')

```

[Fetch boto3 credentials only from EC2 instance profile](https://stackoverflow.com/questions/49496141/fetch-boto3-credentials-only-from-ec2-instance-profile)

**Reply by nsano-rururu** [reply]
I can't say anything because I haven't confirmed it either, but it's possible that the person who added the target implementation hasn't tested it either.
My guess is that the instance profile works fine.

If you are using an instance profile you should have the following code

```
session = boto3.session.Session(region_name=aws_region)
```

If you are using an AWS profile you should have the following code

```
session = boto3.session.Session(profile_name=profile_name)
```

**Reply by nsano-rururu** [reply]
If my hypothesis is correct, it should be the following implementation

```
if profile_name:
    session = boto3.session.Session(profile_name=profile_name)
else:
    session = boto3.session.Session(region_name=aws_region)
```

If I have a chance to use Amazon Elasticsearch service, I will check the operation.

**Reply by jertel** [reply]
I think that would work but possibly add the region to the first branch, for users that want to override their default profile region:
```
if profile_name:
    session = boto3.session.Session(profile_name=profile_name, region_name=aws_region)
else:
    session = boto3.session.Session(region_name=aws_region)
```

**Reply by nsano-rururu** [reply]
At that time, os.environ.get ('AWS_DEFAULT_REGION') is not set in aws_region, so it looks like it is not actually used.

I will report the above questions after creating an environment and moving it.

**Reply by nsano-rururu** [reply]
It has the following specifications. It's definitely overwriting aws_region rather than for users who overwrote aws_region.

■ Specification arrangement
- When using Amazon Elasticsearch service, be sure to write aws_region in config.yaml.
- Profiles are supported, but region values are always overwritten with the values set in config.yaml

**Reply by nsano-rururu** [reply]
As a result of investigating the document description, if the description is different from the implementation, it may be corrected.
https://elastalert2.readthedocs.io/en/latest/recipes/signing_requests.html

**Reply by nsano-rururu** [reply]
Organize


The following statement in the document is a lie. Not implemented. It is assumed that you will use a profile.

https://elastalert2.readthedocs.io/en/latest/recipes/signing_requests.html
>Using an Instance Profile
Typically, you’ll deploy ElastAlert 2 on a running EC2 instance on AWS. You can assign a role to this instance that gives it permissions to read from and write to the Elasticsearch service. When using an Instance Profile, you will need to specify the aws_region in the configuration file or set the AWS_DEFAULT_REGION environment variable.

If the following code is executed when the profile name is not specified with the value of aws_region, the exception `botocore.exceptions.ProfileNotFound: The config profile (a) could not be found` is thrown.

```
session = boto3.session.Session(profile_name=profile_name, region_name=aws_region)
```

The following checks are missing in the current implementation

```
        if not profile_name and not os.environ.get('AWS_DEFAULT_PROFILE'):
            return None
```

---

To see if the following can be achieved, it is necessary to create an environment for Amazon Elasticsearch Service and check the operation.

https://elastalert2.readthedocs.io/en/latest/recipes/signing_requests.html
>Using an Instance Profile
Typically, you’ll deploy ElastAlert 2 on a running EC2 instance on AWS. You can assign a role to this instance that gives it permissions to read from and write to the Elasticsearch service. When using an Instance Profile, you will need to specify the aws_region in the configuration file or set the AWS_DEFAULT_REGION environment variable.

**Reply by jertel** [reply]
Ok. If you are confident it's not implemented feel free to push the fix and doc change. Thanks for your research on this.

**Reply by nsano-rururu** [reply]
There are no plans to make a pull request in a hurry at this time. Because there are no inquiries from users.

**Reply by nsano-rururu** [reply]
The name of `Amazon Elasticsearch Service` was changed to `Amazon OpenSearch Service`
